internal/scheduler/nodedb/nodedb.go

```diff
@@ -607,17 +631,76 @@ func (nodeDb *NodeDb) SelectNodeForJobWithTxn(txn *memdb.Txn, jctx *schedulercon
 		if it, err := txn.Get("nodes", "id", nodeId); err != nil {
 			return nil, errors.WithStack(err)
 		} else {
-			if node, err := nodeDb.selectNodeForPodWithIt(it, jctx, req.Priority, true); err != nil {
+			priority, ok := nodeDb.GetScheduledAtPriority(jctx.JobId)
+			if !ok {
+				// We only get here if the node ID label was set by the user
+				// (instead of the scheduler); home-away preemption is ignored
+				// in that case.
+				priority = req.Priority
+			}
+			pctx.ScheduledAtPriority = priority
+			if node, err := nodeDb.selectNodeForPodWithItAtPriority(it, jctx, priority, true); err != nil {
 				return nil, err
 			} else {
 				return node, nil
 			}
 		}
 	}
 
+	node, err := nodeDb.selectNodeForJobWithTxnAtPriority(txn, jctx, req.Priority)
+	if err != nil {
+		return nil, err
+	}
+	if node != nil {
+		return node, nil
+	}
+
+	// Save the number of additional tolerations that the job originally had; we
+	// use this value to restore the slice of additional toleration at the end
+	// of each loop iteration.
+	numAdditionalTolerations := len(jctx.AdditionalTolerations)
+	for _, awayNodeType := range priorityClass.AwayNodeTypes {
+		wellKnownNodeType, ok := nodeDb.wellKnownNodeTypes[awayNodeType.WellKnownNodeTypeName]
+		if !ok {
+			return nil, fmt.Errorf("unknown well-known node type %s; must be in %v", awayNodeType.WellKnownNodeTypeName, nodeDb.wellKnownNodeTypes)
```

Comment:
We could get jobs into the system with unknown well-known node types, e.g., if we remove a well-known node type after a job is submitted. Do we set the error as the failure reason for the job and then continue scheduling?

Reply:
Do we validate the config to ensure that there are no unknown well-known node types referenced in a PC?

Reply:
Skip this well-known node type if we can't find it and log a warning.